### PR TITLE
Feature/factory structure

### PR DESCRIPTION
**Pull request by ryfurrer**
My understanding is that we have:  
1. modules to define common matrix functionality that cannot come from inheritance,  
2. factories to generate matrices and 
3. different sparse matrix classes to simplify calculations
    - Ex: det for triangular -> product of main diagonal; det for general matrix -> cramer's rule

### CODE REVIEW

File: lib/sparse_matrix.rb [hunk 1]
```diff
@@ -16,38 +16,6 @@ def identity(n)
     alias :I :identity
   end
 
-  def rows
```
Comment (jhryniw) [inline lib/sparse_matrix.rb]: Why are we removing these? They need to be in the interface for the TA to read

File: lib/sparse_matrix_abstract_factory.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Abstract factory for sparse matrices
+class AbstractSparseMatrixFactory
+  @factories = Hash.new
```
Comment (jhryniw) [inline lib/sparse_matrix_abstract_factory.rb]: Doesn't this have to be a class variable (@@)

**Review by jhryniw — DISMISSED** [review]
I'm confused as to the purpose of the member functions in the factory classes

File: lib/sparse_matrix.rb (hunk 1 above)
Comment (ryfurrer) [reply lib/sparse_matrix.rb]: Moved them back to the classes (had them in the factories)

**Comment by ryfurrer** [comment]
The abstract sparse matrix class is for alternate sparse matrix structures like Block Sparse Row matrix (however we are not writing other types).

**Pull request merged**